File: scripts/test-namleh-publish-guards.rb
#!/usr/bin/env ruby

require "json"
require "yaml"

ROOT = File.expand_path("..", __dir__)
UPSTREAM_GUARD = "github.repository == 'block/buzz'"
PUBLISH_PATTERN = /"permission-[a-z-]+":"write"|docker\/build-push-action|\bhelm push\b|\bgh release (?:create|edit|upload)\b|\bgit push\b/

def write_permissions?(permissions)
  return permissions == "write-all" if permissions.is_a?(String)
  return false unless permissions.is_a?(Hash)

  permissions.any? { |_scope, access| access == "write" }
end

def wrapped?(expression)
  return false unless expression.start_with?("(") && expression.end_with?(")")

  depth = 0
  expression.each_char.with_index do |character, index|
    depth += 1 if character == "("
    depth -= 1 if character == ")"
    return false if depth.zero? && index < expression.length - 1
  end
  depth.zero?
end

def exclusive_upstream_guard?(condition)
  normalized = condition&.split&.join(" ")
  return true if normalized == UPSTREAM_GUARD
  return false unless normalized&.start_with?("#{UPSTREAM_GUARD} && ")

  remainder = normalized.delete_prefix("#{UPSTREAM_GUARD} && ").strip
  !remainder.include?("||") || wrapped?(remainder)
end

quoted_fixture = YAML.safe_load(<<~YAML)
  'permissions':
    'contents': 'write'
YAML
raise "quoted permission fixture was not recognized" unless write_permissions?(quoted_fixture.fetch("permissions"))
raise "quoted write-all fixture was not recognized" unless write_permissions?(YAML.safe_load('permissions: "write-all"').fetch("permissions"))

guarded_count = 0
Dir.glob(File.join(ROOT, ".github/workflows/*.{yml,yaml}")).sort.each do |path|
  workflow = YAML.safe_load(File.read(path), aliases: true)
  jobs = workflow.fetch("jobs", {})
  workflow_permissions = workflow["permissions"]

  jobs.each do |job_id, job|
    raise "#{path}: job #{job_id} must be an object" unless job.is_a?(Hash)

    effective_permissions = job.key?("permissions") ? job["permissions"] : workflow_permissions
    serialized = JSON.generate(job)
    dangerous = write_permissions?(effective_permissions) || serialized.match?(PUBLISH_PATTERN)
    next unless dangerous

    unless exclusive_upstream_guard?(job["if"])
      relative = path.delete_prefix("#{ROOT}/")
      raise "#{relative}: write-capable job #{job_id} is not exclusively fork-gated to block/buzz"
    end
    guarded_count += 1
  end
end

raise "no write-capable publication jobs were discovered" if guarded_count.zero?
puts "All #{guarded_count} inherited write-capable jobs are fork-gated."

File: scripts/test-namleh-publish-guards.mjs
import { readFileSync } from "node:fs";
import { resolve } from "node:path";

const root = resolve(import.meta.dirname, "..");
const guardedJobs = {
  ".github/workflows/auto-tag-on-release-pr-merge.yml": ["auto-tag"],
  ".github/workflows/docker.yml": ["build", "merge", "push-gateway-build", "push-gateway-merge"],
  ".github/workflows/helm-chart.yml": ["publish"],
  ".github/workflows/mobile-release-candidate.yml": ["publish"],
  ".github/workflows/promote-oss-desktop-release.yml": ["promote"],
  ".github/workflows/push-gateway-helm-chart.yml": ["publish"],
  ".github/workflows/release.yml": [
    "setup",
    "release",
    "release-macos-x64",
    "release-linux",
    "release-windows",
    "assemble-manifest",
  ],
  ".github/workflows/sprig-image.yml": ["build", "merge"],
  ".github/workflows/sprig.yml": ["publish", "publish-tag"],
};

const forkExclusiveJobs = {
  ".github/workflows/docker.yml": ["build", "push-gateway-build"],
  ".github/workflows/sprig-image.yml": ["build"],
};

for (const [file, jobs] of Object.entries(guardedJobs)) {
  const workflow = readFileSync(resolve(root, file), "utf8");
  for (const job of jobs) {
    const start = workflow.search(new RegExp(`^  ${job}:\\s*$`, "m"));
    if (start < 0) throw new Error(`${file}: missing ${job} job`);
    const remainder = workflow.slice(start + 1);
    const next = remainder.search(/^  [A-Za-z0-9_-]+:\s*$/m);
    const block = next < 0 ? remainder : remainder.slice(0, next);
    if (!block.includes("github.repository == 'block/buzz'")) {
      throw new Error(`${file}: ${job} is not fork-gated to block/buzz`);
    }
  }
}

for (const [file, jobs] of Object.entries(forkExclusiveJobs)) {
  const workflow = readFileSync(resolve(root, file), "utf8");
  for (const job of jobs) {
    const start = workflow.search(new RegExp(`^  ${job}:\\s*$`, "m"));
    const remainder = workflow.slice(start + 1);
    const next = remainder.search(/^  [A-Za-z0-9_-]+:\s*$/m);
    const block = next < 0 ? remainder : remainder.slice(0, next);
    const condition = block.match(/^    if: (.+)$/m)?.[1];
    if (condition !== "github.repository == 'block/buzz'") {
      throw new Error(`${file}: ${job} must be disabled entirely outside block/buzz`);
    }
  }
}

console.log("All inherited publication jobs are fork-gated.");
